File: companies/OpenAI/71.simplify-path.java
import java.util.*;

/**
 * Algorithm:
 * Unix path simplification is stack-based. Directory names push, ".." pops one
 * directory if possible, and "." or empty components are ignored.
 *
 * Java data structures:
 * ArrayDeque<String> acts as a stack of path components.
 *
 * Complexity:
 * Time O(n), space O(n).
 */
class Solution {
    public String simplifyPath(String path) {
        Deque<String> stack = new ArrayDeque<>();
        for (String part : path.split("/")) {
            if (part.isEmpty() || part.equals(".")) {
                continue;
            }
            if (part.equals("..")) {
                if (!stack.isEmpty()) {
                    stack.removeLast();
                }
            } else {
                stack.addLast(part);
            }
        }
        StringBuilder ans = new StringBuilder();
        for (String part : stack) {
            ans.append('/').append(part);
        }
        return ans.length() == 0 ? "/" : ans.toString();
    }
}

File: companies/FPT/71.simplify-path.go
package leetcode

import "strings"

// SimplifyPath71 is stack-based Unix path normalization. Directory names push,
// ".." pops when possible, and "." or empty components from repeated slashes do
// nothing. Joining the stack yields the canonical absolute path.
//
// Time: O(n). Space: O(n).
func SimplifyPath71(path string) string {
	stack := []string{}
	for _, part := range strings.Split(path, "/") {
		if part == "" || part == "." {
			continue
		}
		if part == ".." {
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		} else {
			stack = append(stack, part)
		}
	}
	return "/" + strings.Join(stack, "/")
}
